src/freenas/usr/local/bin/truenas-initrd.py

```diff
@@ -12,40 +13,109 @@
+
+    if restore_state is not None:
+        mountpoints_mapping = {
+            k.split("_")[-1]: v for k, v in readonly_state.__dict__.items() if k.startswith("original_")
+        }
+    else:
+        mountpoints_mapping = {
+            "root": readonly_desired_value,
+            "usr": readonly_desired_value,
+        }
+
+    for key, value in filter(lambda i: i[0].startswith("current_"), readonly_state.__dict__.items()):
+        key = key.split("_")[-1]
+        if mountpoints_mapping[key] == value:
+            mountpoints_mapping.pop(key)
+
+    if not mountpoints_mapping:
         return
 
     # Some initramfs scripts use (`dpkg --print-architecture` or similar calls)
-    if readonly:
+    if mountpoints_mapping.get("usr") is True:
         os.chmod(os.path.join(root, "usr/bin/dpkg"), 0o644)
-        os.rename(os.path.join(root, "usr/local/bin/dpkg.bak"), os.path.join(root, "usr/local/bin/dpkg"))
-
-    readonly_value = "on" if readonly else "off"
-    mountpoints = [root, os.path.join(root, "usr")]
-    for partition in psutil.disk_partitions():
-        if partition.mountpoint in mountpoints and partition.fstype == "zfs":
-            # Do not change `readonly` property when we're running in the installer, and it was not set yet
-            if (
-                subprocess.run(
-                    ["zfs", "get", "-H", "-o", "source", "readonly", partition.device],
-                    capture_output=True, text=True,
-                ).stdout.strip() == "local"
-            ):
-                subprocess.run(["zfs", "set", f"readonly={readonly_value}", partition.device])
-
-    if not readonly:
-        os.chmod(os.path.join(root, "usr/bin/dpkg"), 0o755)
-        os.rename(os.path.join(root, "usr/local/bin/dpkg"), os.path.join(root, "usr/local/bin/dpkg.bak"))
+        with contextlib.suppress(FileNotFoundError):
+            # In the event this is called first with readonly set
+            os.rename(os.path.join(root, "usr/local/bin/dpkg.bak"), os.path.join(root, "usr/local/bin/dpkg"))
+
+    for device_key, ds_name in filter(lambda i: i[0] in mountpoints_mapping, get_ds_conf(root).items()):
+        # Do not change `readonly` property when we're running in the installer, and it was not set yet
+        if (
+            subprocess.run(
+                ["zfs", "get", "-H", "-o", "source", "readonly", ds_name],
+                capture_output=True, text=True,
+            ).stdout.strip() == "local"
+        ):
+            readonly_value = mountpoints_mapping[device_key]
+            subprocess.run(["zfs", "set", f"readonly={'on' if readonly_value else 'off'}", ds_name])
+            setattr(readonly_state, f"current_{device_key}", readonly_value)
 
-    readonly_state = readonly
+    if mountpoints_mapping.get("usr") is False:
+        os.chmod(os.path.join(root, "usr/bin/dpkg"), 0o755)
+        with contextlib.suppress(FileNotFoundError):
+            # In the event this is called twice with readonly unset
```

Comment:
I think we should expand the comment explaining what `/usr/local/bin/dpkg` is and why we are renaming it.